capirca/lib/fortigate.py

```diff
@@ -0,0 +1,330 @@
+
+    supported_tokens |= {'source_interface',
+                         'destination_interface',
+                         'logging'}
+
+    supported_sub_tokens.update({'option': {'from_id'},
+                                 # Warning, some of these are mapped
+                                 # differently. See _ACTION_TABLE
+                                 'action': {'accept', 'deny', 'reject',
+                                            'reject-with-tcp-rst'}})
+    return supported_tokens, supported_sub_tokens
+
+  def _TranslatePolicy(self, pol, exp_info):
+    self.fortigate_policies = []
+    current_date = datetime.datetime.utcnow().date()
+    exp_info_date = current_date + datetime.timedelta(weeks=exp_info)
+
+    # a mixed filter outputs both ipv4 and ipv6 acls in the same output file
+
+    for header, terms in pol.filters:
+      if self._PLATFORM not in header.platforms:
+        continue
+
+      filter_options = header.FilterOptions(self._PLATFORM)
+
+      if (len(filter_options) < 2 or filter_options[0] != "from-id"):
+        raise UnsupportedFilterError(
+          "Fortigate Firewall filter arguments must specify from_id")
+
+      from_id = filter_options[1]
+      Term.CURRENT_ID = int(from_id)
+
+      self.verbose = True
+      if 'noverbose' in filter_options:
+        filter_options.remove('noverbose')
+        self.verbose = False
+
+      term_dup_check = set()
+
+      for term in terms:
+        filter_name = header.FilterName(self._PLATFORM)
+        if term.expiration:
+          if term.expiration <= exp_info_date:
+            logging.info('INFO: Term %s in policy %s expires '
+                         'in less than two weeks.', term.name, filter_name)
+          if term.expiration <= current_date:
+            logging.warn('WARNING: Term %s in policy %s is expired and '
+                         'will not be rendered.', term.name, filter_name)
+            continue
+        if term.name in term_dup_check:
+          raise FortiGateDuplicateTermError('You have a duplicate term: %s' %
+                                            term.name)
+        term_dup_check.add(term.name)
+
+        term.name = self.FixTermLength(term.name)
+        new_term = Term(term)
+
+        self.fortigate_policies.append((header, term.name, new_term))
+
+  def _GetTargetByPolicyID(self, id):
+    return '\tedit %s' % id
+
+  def __str__(self):
+    start_addresses = ['config firewall address']
+    start_services = ['config firewall service custom']
+    start_policies = ['config firewall policy']
+    end = ['end']
+    target_addresses = []
+    target_services = []
+    target_policies = []
+
+    for (header, filter_name, term) in self.fortigate_policies:
+      target_policies.append(self._GetTargetByPolicyID(term.id))
+
+      term_str = str(term)
+
+      target_policies.append(term_str)
+
+      target_policies += ['\tnext', '']
+    target_addresses.extend(Term.get_fw_addresses())
```

Comment:
I am unsure if this is a good idea to be using a static method like this for this purpose. You are already looping over the terms so collecting the addresses shouldn't be a problem. Storing a list of all the IP's in the class is also weird and feels incorrect to me.

Reply:
for same reason explained here:
 https://github.com/google/capirca/pull/173#discussion_r292202610